pkg/3scale/amp/operator/upgrade.go

```diff
@@ -394,6 +405,124 @@ func (u *UpgradeApiManager) findDeploymentTriggerOnImageChange(triggerPolicies [
+	return reconcile.Result{}, nil
+}
+
+func (u *UpgradeApiManager) upgradeSystemAppBackendRouteEnv(desired *appsv1.DeploymentConfig) (reconcile.Result, error) {
+	existing := &appsv1.DeploymentConfig{}
+	err := u.Client().Get(context.TODO(), types.NamespacedName{Name: desired.Name, Namespace: u.apiManager.Namespace}, existing)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if len(existing.Spec.Template.Spec.Containers) != 3 {
+		return reconcile.Result{}, fmt.Errorf("DeploymentConfig %s spec.template.spec.containers length is %d, should be 3",
+			existing.Name, len(existing.Spec.Template.Spec.Containers))
+	}
+
+	desiredName := common.ObjectInfo(desired)
+
+	update := false
+	for idx := 0; idx < 3; idx++ {
+		existingContainer := &existing.Spec.Template.Spec.Containers[idx]
+		desiredContainer := &desired.Spec.Template.Spec.Containers[idx]
+		desiredBackendRouteEnvVar, ok := helper.FindEnvVar(desiredContainer.Env, "BACKEND_ROUTE")
+		if !ok {
+			return reconcile.Result{}, fmt.Errorf("%s desired spec.template.spec.containers env var '%s' does not exist", desiredName, "BACKEND_ROUTE")
+		}
+		tmpUpdate := ensureBackendRouteEnvVar(desiredBackendRouteEnvVar, &existingContainer.Env)
+		update = update || tmpUpdate
+	}
+
+	// Pre hook pod env vars
+	desiredBackendRouteEnvVar, ok := helper.FindEnvVar(desired.Spec.Strategy.RollingParams.Pre.ExecNewPod.Env, "BACKEND_ROUTE")
+	if !ok {
+		return reconcile.Result{}, fmt.Errorf("%s desired spec.strategy.rollingparams.pre.execnewpod env var '%s' does not exist", desiredName, "BACKEND_ROUTE")
+	}
+	tmpUpdate := ensureBackendRouteEnvVar(desiredBackendRouteEnvVar, &existing.Spec.Strategy.RollingParams.Pre.ExecNewPod.Env)
+	update = update || tmpUpdate
+
+	if update {
+		u.Logger().Info(fmt.Sprintf("Upgrading BACKEND_ROUTE environment variable to DC %s", existing.Name))
+		err = u.UpdateResource(existing)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{Requeue: update}, nil
+}
+
+func (u *UpgradeApiManager) upgradeSidekiqBackendRouteEnv(desired *appsv1.DeploymentConfig) (reconcile.Result, error) {
+	existing := &appsv1.DeploymentConfig{}
+	err := u.Client().Get(context.TODO(), types.NamespacedName{Name: desired.Name, Namespace: u.apiManager.Namespace}, existing)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if len(existing.Spec.Template.Spec.Containers) != 1 {
+		return reconcile.Result{}, fmt.Errorf("DeploymentConfig %s spec.template.spec.containers length is %d, should be 1",
+			existing.Name, len(existing.Spec.Template.Spec.Containers))
+	}
+
+	desiredName := common.ObjectInfo(desired)
+
+	existingContainer := &existing.Spec.Template.Spec.Containers[0]
+	desiredContainer := &desired.Spec.Template.Spec.Containers[0]
+	desiredBackendRouteEnvVar, ok := helper.FindEnvVar(desiredContainer.Env, "BACKEND_ROUTE")
+	if !ok {
+		panic(fmt.Sprintf("%s desired spec.template.spec.containers env var '%s' does not exist", desiredName, "BACKEND_ROUTE"))
```

Comment:
We should apply the same criteria as with the other, just return an error

Reply:
done